misk-aws/src/main/kotlin/misk/jobqueue/sqs/MiskJobModule.kt

```diff
@@ -0,0 +1,124 @@
+import misk.inject.KAbstractModule
+import misk.inject.asSingleton
+import misk.inject.toKey
+import misk.jobqueue.JobConsumer
+import misk.jobqueue.JobHandler
+import misk.jobqueue.JobQueue
+import misk.jobqueue.QueueName
+import java.time.Duration
+import javax.inject.Provider
+import kotlin.reflect.KClass
+
+/**
+ * Install multiple job queues backed by AwsSQS.
+ *
+ * Usage
+ *
+ * In your misk service Module include:
+ *  ```
+ *  install(MiskJobModule(environment)
+ *    .withHandler<SomeJobHandlerA>(JobQueueA::class, "job_queue_a")
+ *    .withHandler<SomeJobHandlerB>(JobQueueB::class, "job_queue_b"))
+ *
+ *  ```
+ *  and later access using:
+ *  ```
+ *  @Inject @JobQueueA lateinit var jobQueueA: MiskJobQueue`
+ *  ```
+ */
+
+class MiskJobModule(private val environment: Environment) : KAbstractModule() {
+  private lateinit var jobQueueProvider: Provider<JobQueue>
+  private val jobQueuesMap =
+      mutableListOf<JobQueueBinding>()
+
+  inline fun <reified T : JobHandler> withHandler(
+    qualifier: KClass<out Annotation>, queueName: String
+  ) = withHandler(qualifier, queueName, T::class)
+
+  fun withHandler(
+    qualifier: KClass<out Annotation>,
+    queueName: String,
+    jobHandler: KClass<out JobHandler>
+  ): MiskJobModule {
+    jobQueuesMap += JobQueueBinding(qualifier,
+        QueueName(queueName),
+        jobHandler)
+    return this
+  }
+
+  private fun configureJobQueues() {
+    for ((qualifier, queueName, _) in jobQueuesMap) {
+      val queueKey = MiskJobQueue::class.toKey(qualifier)
+
+      bind(queueKey).toProvider(Provider<MiskJobQueue> {
+        object : MiskJobQueue {
+          override fun enqueue(
+            body: String,
+            deliveryDelay: Duration?,
+            attributes: Map<String, String>
+          ) {
+            jobQueueProvider.get().enqueue(queueName, body, deliveryDelay, attributes)
+          }
+        }
+      }).asSingleton()
+    }
+  }
+
+  private fun bindHandlers(): Map<QueueName, Provider<JobHandler>> =
+      jobQueuesMap.map { (qualifier, queueName, jobHandler) ->
+        val jobHandlerKey = JobHandler::class.toKey(qualifier)
+        val jobHandlerProvider = getProvider(jobHandlerKey)
+        bind(jobHandlerKey).to(jobHandler.java)
+        queueName to jobHandlerProvider
+      }.toMap()
+
+  override fun configure() {
+    val jobQueueKey = JobQueue::class.toKey()
+    jobQueueProvider = getProvider(jobQueueKey)
+    val jobConsumerKey = JobConsumer::class.toKey()
+    val jobConsumerProvider = getProvider(jobConsumerKey)
+    val jobQueueServiceKey = MiskJobService::class.toKey()
+
+    configureJobQueues()
+    val boundHandlers = bindHandlers()
+
+    bind(jobQueueServiceKey).toProvider(Provider<MiskJobService> {
+      MiskJobService(jobConsumerProvider.get(), boundHandlers)
+    }).asSingleton()
+
+    multibind<Service>().to(jobQueueServiceKey)
+
+    if (environment == Environment.DEVELOPMENT || environment == Environment.TESTING) {
```

Comment:
This will prevent applications from installing multiple `MiskJobQueueModule`s, which in turn prevents you from using `MiskJobQueue` and job consumer auto-registration from within components that might get installed by the application. The pattern we've generally taken in misk is to have one modules for global singletons which gets installed by the application, and another module for registering "handlers" that get used by those global singletons.  That way a component provider can provide a top-level module that installs the component specific handlers, without it also installing duplicate bindings for the global singletons.

cf. `MiskWebModule`